app/components/autocomplete/channel_mention/channel_mention.tsx

```diff
@@ -0,0 +1,308 @@
+
+        if (Platform.OS === 'ios') {
+            // This is the second part of the hack were we replace the double ~ with just one
+            // after the auto correct vanished
+            setTimeout(() => {
+                updateValue(completedDraft.replace(`~~${mention} `, `~${mention} `));
+            });
+        }
+
+        onShowingChange(false);
+        setNoResultsTerm(mention);
+        setSections([]);
+    }, [value, localCursorPosition, isSearch]);
+
+    const renderItem = useCallback(({item}) => {
+        return (
+            <ChannelMentionItem
+                channel={item}
+                onPress={completeMention}
+                testID={`autocomplete.channel_mention.item.${item}`}
+            />
+        );
+    }, [completeMention]);
+
+    const renderSectionHeader = useCallback(({section}) => {
+        return (
+            <AutocompleteSectionHeader
+                id={section.id}
+                defaultMessage={section.defaultMessage}
+                loading={!section.hideLoadingIndicator && loading}
+            />
+        );
+    }, [loading]);
+
+    useEffect(() => {
+        if (localCursorPosition !== cursorPosition) {
+            setLocalCursorPosition(cursorPosition);
+        }
+    }, [cursorPosition]);
+
+    useEffect(() => {
+        if (matchTerm === null) {
+            resetState();
+            onShowingChange(false);
+            return;
+        }
+
+        if (noResultsTerm != null && matchTerm.startsWith(noResultsTerm)) {
+            return;
+        }
+
+        setNoResultsTerm(null);
+        runSearch(serverUrl, matchTerm);
+    }, [matchTerm]);
+
+    useDidUpdate(() => {
+        const newSections = [];
+        if (isSearch) {
+            const [publicChannels, privateChannels, directAndGroupMessages] = reduceChannelsForSearch(channels, myMembers);
+            if (publicChannels.length) {
+                newSections.push({
+                    id: t('suggestion.search.public'),
+                    defaultMessage: 'Public Channels',
+                    data: publicChannels,
+                    key: 'publicChannels',
+                    hideLoadingIndicator: true,
+                });
+            }
+
+            if (privateChannels.length) {
+                newSections.push({
+                    id: t('suggestion.search.private'),
+                    defaultMessage: 'Private Channels',
+                    data: privateChannels,
+                    key: 'privateChannels',
+                    hideLoadingIndicator: true,
+                });
+            }
+
+            if (directAndGroupMessages.length) {
+                newSections.push({
+                    id: t('suggestion.search.direct'),
+                    defaultMessage: 'Direct Messages',
+                    data: directAndGroupMessages,
+                    key: 'directAndGroupMessages',
+                    hideLoadingIndicator: true,
+                });
+            }
```

Comment:
I like the pure functions you placed outside of the component and at the top of the file for `reduceChannelsForSearch` and `reduceChannelsForAutocomplete`. 1/5 but you could also do the same for these section and reduce this to something like `getSearchSections()`.  That same could apply for the `!isSearch` `else` block below.